### PR TITLE
chore: track terraform modules in telemetry

### PR DESCRIPTION
**Pull request by hugodutka**
Addresses https://github.com/coder/nexus/issues/35.

This PR:

- Adds a `workspace_modules` table to track modules used by the Terraform provisioner in provisioner jobs.
- Adds a `module_path` column to the `workspace_resources` table, allowing to identify which module a resource originates from.
- Starts pushing this new information into telemetry.

For the person reviewing this PR, do not fret about the 1,500 new lines - ~1,000 of them are auto-generated.

### CODE REVIEW

**Comment by bpmct** [comment]
wooo!!!

File: provisioner/terraform/resources.go
Comment (johnstcn) [inline provisioner/terraform/resources.go:603]: Could we instead drop an error log? This feels very brittle to me.

File: coderd/database/migrations/000274_workspace_modules.up.sql
Comment (johnstcn) [inline coderd/database/migrations/000274_workspace_modules.up.sql:14]: Why do we need to insert these to the database? Is there any other reason we might want to save this data, or is it only for telemetry?

File: provisioner/terraform/resources.go
Comment (johnstcn) [inline provisioner/terraform/resources.go:774]:
```suggestion
		return "", xerrors.Errorf("parse terraform module address: %w", err)
```

File: provisioner/terraform/provision.go
Comment (johnstcn) [inline provisioner/terraform/provision.go:148]: So, if we fail to get the list of modules for the build that we _only need_ for telemetry purposes, we fail to build?

File: coderd/database/migrations/000274_workspace_modules.up.sql
Comment (hugodutka) [reply coderd/database/migrations/000274_workspace_modules.up.sql:14]: This is only for telemetry purposes. Since `CompleteJob` reports telemetry after finishing, we could skip saving the data in the database. The only downside is that if the `CompleteJob` report fails for any reason, we won't be able to retry sending it later.

I don't have a strong opinion on this - if you prefer we skipped saving, I'll be happy to do that.

File: provisioner/terraform/provision.go
Comment (hugodutka) [reply provisioner/terraform/provision.go:148]: Good catch, it'd probably be better to only log an error.

File: provisioner/terraform/resources.go
Comment (hugodutka) [reply provisioner/terraform/resources.go:603]: We could. However, leaving the module path as an empty string would imply it's the "root module," which isn't our intention. I wanted to avoid making the module path nullable here because, in the database, a null value in `WorkspaceResource`'s `ModulePath` indicates "this resource was created before module paths were tracked." Do you think setting it to null here would be a better approach?

File: coderd/database/dump.sql
Comment (johnstcn) [inline coderd/database/dump.sql:1695]: 🤔 Why does the `workspace_proxies` table have a `module_path` and nothing else?

File: coderd/database/migrations/000274_workspace_modules.up.sql
Comment (dannykopping) [inline coderd/database/migrations/000274_workspace_modules.up.sql:9]: Do we need to capture the module for all transitions? What's the value of this?

File: provisioner/terraform/resources.go
Comment (johnstcn) [reply provisioner/terraform/resources.go:603]: My personal preference would be to leave it as a null string in order to avoid conflating it with "root module". We could possibly have a static sentinel value for "invalid module path" either if there is a need to differentiate between "invalid module path" and "not tracked".

File: coderd/database/migrations/000274_workspace_modules.up.sql
Comment (dannykopping) [reply coderd/database/migrations/000274_workspace_modules.up.sql:14]: This also had me scratching my head.

> The only downside is that if the CompleteJob report fails for any reason, we won't be able to retry sending it later.

_Do_ we retry if telemetry fails to send?

File: provisioner/terraform/modules.go
Comment (dannykopping) [inline provisioner/terraform/modules.go:1]: I'm curious about this.

We extract all other resources from the template by executing `terraform graph` (see `provisioner/terraform/executor.go` -> `graph()`) and then pass that to `provisioner/terraform/resources.go` -> `ConvertState()`. From there we retrieve all the template's resources, and they're then persisted.

Did this approach not work for modules?

File: coderd/database/migrations/000274_workspace_modules.up.sql
Comment (dannykopping) [inline coderd/database/migrations/000274_workspace_modules.up.sql:1]: I'm curious why you opted for a new table rather than using `workspace_resources`?

File: coderd/database/migrations/000274_workspace_modules.up.sql
Comment (johnstcn) [reply coderd/database/migrations/000274_workspace_modules.up.sql:14]: This is my first time really delving into the telemetry stuff, but here's what I think is true at the time of writing (please correct me if I'm wrong):

1) Telemetry snapshots are done periodically so any data we do not persist will be lost between coderd restarts
2) Telemetry snapshots are done by each coderd independently, so data that is known to one `coderd` only may not be present in snapshots from other coderd instances.

I'm not sure if there is any sort of de-duplication logic on the telemetry server end, but having to merge multiple disparate snapshots sounds like it would get messy and error-prone.

All of that is basically to say: we probably do need to store in the database in order to get a consistent snapshot.

**Comment by johnstcn** [comment]
Thought: I can imagine a valid use-case for storing module data separate to telemetry -- showing in the UI which module a resource came from is probably useful from a template editor / admin perspective when trying to answer the question "Which module did this resource come from?"

File: coderd/database/migrations/000274_workspace_modules.up.sql
Comment (johnstcn) [reply coderd/database/migrations/000274_workspace_modules.up.sql:1]: Given that you can have multiple versions of a module, having all that data in `workspace_resources` would likely result in a lot of excess data storage.

Comment (dannykopping) [reply coderd/database/migrations/000274_workspace_modules.up.sql:1]: I don't see how having it in this separate table would improve that?

Comment (johnstcn) [reply coderd/database/migrations/000274_workspace_modules.up.sql:1]: Yeah I looked at the proposed schema more closely and noticed that each module is linked to a specific provisioner job.

I think there is an argument to be made that each version of a given module should only be stored once in the database. That same module version can then be referenced by multiple workspace resources presumably at a different module path?

File: coderd/provisionerdserver/provisionerdserver_test.go
Comment (johnstcn) [inline coderd/provisionerdserver/provisionerdserver_test.go:1522]: Will this also track private modules? If so, should we, and/or do we care about private moudle usage?

File: coderd/database/migrations/000274_workspace_modules.up.sql
Comment (hugodutka) [reply coderd/database/migrations/000274_workspace_modules.up.sql:14]: @dannykopping we run a periodic snapshot of a subset of tables in the db created [within the last hour](https://github.com/coder/coder/blob/ea2b6b7cbc9022ec92ee49f74c75aefd77613360/coderd/telemetry/telemetry.go#L292). So we actually store multiple instances of rows with the same id in our telemetry db and deduplicate when querying the data. That's what I meant by retrying - if the `CompleteJob` snapshot fails to send, the next periodic snapshot will send the same data again.

@johnstcn as [discussed on Slack](https://codercom.slack.com/archives/CJURPL8DN/p1731322612970969) storing the data in PG may not be required for consistency, but storing the module data will likely come in handy elsewhere, so it's worth doing anyway.

File: provisioner/terraform/modules.go
Comment (hugodutka) [reply provisioner/terraform/modules.go:1]: I couldn't find either module versions or their sources inside the state. This was the first approach I tried, and I switched to reading files only when I couldn't make it work.

Comment (dannykopping) [reply provisioner/terraform/modules.go:1]: This is the output from a `terraform graph` run locally with a module specified:

```bash
$ terraform graph
digraph G {
  ...
  "module.jetbrains_gateway.coder_app.gateway" -> "coder_agent.main";
  "module.jetbrains_gateway.coder_app.gateway" -> "module.jetbrains_gateway.data.coder_parameter.jetbrains_ide";
  "module.jetbrains_gateway.coder_app.gateway" -> "module.jetbrains_gateway.data.coder_workspace.me";
  "module.jetbrains_gateway.coder_app.gateway" -> "module.jetbrains_gateway.data.coder_workspace_owner.me";
  "module.jetbrains_gateway.coder_app.gateway" -> "module.jetbrains_gateway.data.http.jetbrains_ide_versions";
}
```

File: coderd/database/migrations/000274_workspace_modules.up.sql
Comment (hugodutka) [reply coderd/database/migrations/000274_workspace_modules.up.sql:1]: We don't store all terraform resources in the `workspace_resources` table. We [filter out](https://github.com/coder/coder/blob/ea2b6b7cbc9022ec92ee49f74c75aefd77613360/provisioner/terraform/resources.go#L594) everything Coder related. So to track modules like [personalize](https://registry.coder.com/modules/personalize) which only create a `coder_script` resource, we can't rely on just the resources table.

File: coderd/database/migrations/000274_workspace_modules.up.sql
Comment (hugodutka) [reply coderd/database/migrations/000274_workspace_modules.up.sql:9]: I [mirrored](https://github.com/coder/coder/blob/ea2b6b7cbc9022ec92ee49f74c75aefd77613360/coderd/database/dump.sql#L1688) `workspace_resources`. I think for template import jobs you get resources that reference the same provisioner job but have different workspace transitions. My thinking was if you'd like to associate a specific resource from a template import job with its module, you'd need to know which transition it came from.

File: provisioner/terraform/provision.go
Comment (hugodutka) [reply provisioner/terraform/provision.go:148]: On second thought, after the discussion regarding using modules in places other than telemetry, it may make sense to return an error rather than just log. Otherwise any code using module info would have to take into account that it may be missing.

File: coderd/database/migrations/000274_workspace_modules.up.sql
Comment (hugodutka) [reply coderd/database/migrations/000274_workspace_modules.up.sql:14]: @dannykopping Do you mean storing the provider info in the database, telemetry, or both? This is likely out of scope for this PR - would you like to create an issue for it?

File: provisioner/terraform/modules.go
Comment (hugodutka) [reply provisioner/terraform/modules.go:1]: Yeah, the graph only displays the module names specified by the user in the template. It doesn’t show the source, like `registry.coder.com/modules/jetbrains-gateway/coder`, or the version, like `1.0.23`. That information is available in the files on disk.

**Comment by hugodutka** [comment]
@johnstcn I addressed your comments:

- changed the invalid module path string to a static sentinel string
- obfuscated module source and version for modules outside of `registry.coder.com`
- changed the error message when we fail to parse a terraform address

**Review by johnstcn — COMMENTED** [review]
I have a few more comments and issues I'd like to see clarified before merging this.

File: coderd/database/migrations/000274_workspace_modules.up.sql
Comment (johnstcn) [inline coderd/database/migrations/000274_workspace_modules.up.sql:8]: pair review (non-blocking): alternatively we could normalize this and instead have a `workspace_resources.module_id` column reference an individual row of the `workspace_modules` table so we don't end up storing multiple instances of the same module.

File: coderd/provisionerdserver/provisionerdserver.go
Comment (johnstcn) [reply coderd/provisionerdserver/provisionerdserver.go:1272]: pair review: why do we need start modules _and_ stop modules? What is the difference between the two?

File: coderd/provisionerdserver/provisionerdserver.go
Comment (johnstcn) [inline coderd/provisionerdserver/provisionerdserver.go:1806]: Do we explicitly skip over the root module when we extract it from `modules.json`? IIRC `tfaddr.NewAddress` does not consider an empty address valid? 

Do we even _care_ about the root module?

File: coderd/database/queries/workspacemodules.sql
Comment (johnstcn) [inline coderd/database/queries/workspacemodules.sql:16]: non-blocking: If we're going to be querying on `created_at`, do we need an index?

File: coderd/database/migrations/testdata/fixtures/000274_workspace_modules.up.sql
Comment (johnstcn) [inline coderd/database/migrations/testdata/fixtures/000274_workspace_modules.up.sql:1]: ⚠️ You'll need to rename the migration number. I think you may want 276, but check just before you merge.

File: coderd/database/migrations/000274_workspace_modules.up.sql
Comment (dannykopping) [reply coderd/database/migrations/000274_workspace_modules.up.sql:14]: Sorry for muddying the waters, let's not worry about providers for now.

**Review by johnstcn — CHANGES_REQUESTED** [review]
Requesting changes based on pair review with @dannykopping.

The major worry we have is that there is the _possibility_ that a failure to read the modules on-disk will result in a failure to `Plan()`. While a malformed or otherwise problematic `modules.json` will _probably_ cause Terraform to bail, we shouldn't.

File: coderd/telemetry/telemetry_test.go
Comment (hugodutka) [reply coderd/telemetry/telemetry_test.go:144]: @johnstcn, are you suggesting that instead of obfuscating, we should not report modules outside of registry.coder.com at all?

Comment (johnstcn) [reply coderd/telemetry/telemetry_test.go:144]: What possible use could we have for this sort of data? It's opaque to us. At most, we could report just "internal" and leave it at that.

Comment (hugodutka) [reply coderd/telemetry/telemetry_test.go:144]: The one that comes to my mind is general analytics - we could see that some, likely public, modules are used across many deployments.

File: coderd/database/queries/workspacemodules.sql
Comment (hugodutka) [reply coderd/database/queries/workspacemodules.sql:16]: Won't hurt, I'll add it. 😄

File: coderd/database/migrations/000274_workspace_modules.up.sql
Comment (johnstcn) [reply coderd/database/migrations/000274_workspace_modules.up.sql:8]: That's completely fair 👍 Thanks for clarifying!

File: provisioner/terraform/provision_test.go
Comment (hugodutka) [reply provisioner/terraform/provision_test.go:92]: Tests pass on Windows, so it doesn't seem to matter.

Comment (johnstcn) [reply provisioner/terraform/provision_test.go:92]: 👍 Sounds like `archive/tar` handles it for us!

File: coderd/provisionerdserver/provisionerdserver.go
Comment (hugodutka) [reply coderd/provisionerdserver/provisionerdserver.go:1272]: > note: these will be inserted in random order; is this an issue?

There are no dependencies between modules in the db, so it shouldn't be.

> pair review: why do we need start modules and stop modules? What is the difference between the two?

I think I [answered this before](https://github.com/coder/coder/pull/15450#discussion_r1836591517).

File: coderd/provisionerdserver/provisionerdserver.go
Comment (hugodutka) [reply coderd/provisionerdserver/provisionerdserver.go:1806]: We do skip the root module [when we extract it](https://github.com/coder/coder/blob/b72b8ef3545e43dffbb8417d8d040818177e19d7/provisioner/terraform/modules.go#L46). We don't care about it, as it's always present.

I didn't check if `tfaddr.NewAddress` considers an empty string a valid address, but that shouldn't matter here. In normal operation [`convertAddressToModulePath`](https://github.com/coder/coder/blob/b72b8ef3545e43dffbb8417d8d040818177e19d7/provisioner/terraform/resources.go#L781) should never receive an empty string as input, as resources' addresses are non-empty. It can produce an empty string as output, and `ModulePath` here is that output.

File: provisioner/terraform/provision.go
Comment (hugodutka) [reply provisioner/terraform/provision.go:148]: > Are there any other possible scenarios I'm missing?

I don't think there are any other scenarios - barring disk read failures and such. In normal operation this should only happen if terraform changes how it stores modules on disk and we upgrade to that version.

File: provisionerd/runner/runner.go
Comment (johnstcn) [reply provisionerd/runner/runner.go:1041]: Thanks, a comment would definitely help. From someone else's perspective, this simply looks like a typo.

File: provisioner/terraform/executor.go
Comment (hugodutka) [reply provisioner/terraform/executor.go:489]: Hmm, could you explain? It seems the information required for the log is only available within `ConvertState`.

Comment (johnstcn) [reply provisioner/terraform/executor.go:489]: My explanation is that I had not had enough coffee 🤦

**Comment by hugodutka** [comment]
@johnstcn @dannykopping I made changes based on your feedback.

### List of changes

Each point references the source comment.

- [rebase and update migration number](https://github.com/coder/coder/pull/15450#discussion_r1843547844)
- [add an index on workspace_modules created_at](https://github.com/coder/coder/pull/15450#discussion_r1843543118)
- add a comment why we use `planComplete.Modules` instead of `applyComplete.Modules`. [Comment](https://github.com/coder/coder/pull/15450#discussion_r1842453974).
- [convert sentinel string to error](https://github.com/coder/coder/pull/15450#discussion_r1842455850)
    - added a test for it
- [don’t fail if we can’t get modules from disk](https://github.com/coder/coder/pull/15450#discussion_r1843489464)
- [add a test with a malformed or otherwise invalid module](https://github.com/coder/coder/pull/15450#discussion_r1842472265)
- [combine assignment and conditional](https://github.com/coder/coder/pull/15450#discussion_r1842869899)
- [`dbtestutil.NewDB` instead of dbmem](https://github.com/coder/coder/pull/15450#discussion_r1843493733)
- [extract IsCoderModule to a top-level function](https://github.com/coder/coder/pull/15450#discussion_r1843517113)
- [add a follow up issue about RBAC for workspace_modules](https://github.com/coder/coder/pull/15450#discussion_r1842872282)
    - https://github.com/coder/coder/issues/15543

File: provisioner/terraform/provision_test.go
Comment (johnstcn) [inline provisioner/terraform/provision_test.go:98]: I found many instances of this function sprinkled around our codebase so I added a helper function for this in https://github.com/coder/coder/pull/15540

I'll wait for this to be merged and then update it to use `testutil.CreateTar` :+1:

**Review by johnstcn — APPROVED** [review]
Thanks for working on this!

**Pull request merged**
